### PR TITLE
Address static analysis warnings

### PR DESCRIPTION
**Pull request by yuslepukhin**
**Description**:
Address static analysis warnings.

**Motivation and Context**
Do the right thing.

### CODE REVIEW

File: onnxruntime/python/onnxruntime_pybind_mlvalue.cc [hunk 1]
```diff
@@ -160,14 +160,16 @@ AllocatorPtr GetCudaAllocator(OrtDevice::DeviceId id) {
   // Current approach is not thread-safe, but there are some bigger infra pieces to put together in order to make
   // multi-threaded CUDA allocation work we need to maintain a per-thread CUDA allocator
 
-  static auto* id_to_allocator_map = new std::unordered_map<OrtDevice::DeviceId, AllocatorPtr>();
+  static std::unordered_map<OrtDevice::DeviceId, AllocatorPtr> id_to_allocator_map;
```
Comment (snnn) [inline onnxruntime/python/onnxruntime_pybind_mlvalue.cc:163]: I don't understand the comment. How does it become to per-thread? This is a thread-local variable, and the key is device id, not thread it.

Comment (snnn) [reply onnxruntime/python/onnxruntime_pybind_mlvalue.cc:163]: A big difference between the new old and the old code is that in the new old the destructor of an allocator could be called after the EPs get destroyed.  So, would an allocator hold any pointer to an EP instance?  Though in general we should not use raw new/delete and should avoid memory leak, but sometimes memory leaks is acceptable when destruction order is hard to maintain.

Comment (yuslepukhin) [reply onnxruntime/python/onnxruntime_pybind_mlvalue.cc:163]: I am not the author, just moved the code around some time ago. It is referring to a per-thread allocators, but the map is not per-thread.

Comment (yuslepukhin) [reply onnxruntime/python/onnxruntime_pybind_mlvalue.cc:163]: > A big difference between the new old and the old code is that in the new old the destructor of an allocator could be called after the EPs get destroyed. So, would an allocator hold any pointer to an EP instance? Though in general we should not use raw new/delete and should avoid memory leak, but sometimes memory leaks is acceptable when destruction order is hard to maintain.

I think, as long as it is not used at the time of destruction, it should be fine. This function hands out pointers to the allocators, so anyone can still hold it.

Comment (yuslepukhin) [reply onnxruntime/python/onnxruntime_pybind_mlvalue.cc:163]: Would it be Okay, if those pointers were not owning?

Comment (snnn) [reply onnxruntime/python/onnxruntime_pybind_mlvalue.cc:163]: AllocatorPtr is a shared_ptr. It could be the last one with an ownership.

Comment (yuslepukhin) [reply onnxruntime/python/onnxruntime_pybind_mlvalue.cc:163]: I can make it non-owining

Comment (snnn) [reply onnxruntime/python/onnxruntime_pybind_mlvalue.cc:163]: I think a raw PTR would be better.

**Review by snnn — APPROVED** [review]

**Pull request merged**